src/auto_splitting/mod.rs

```diff
@@ -0,0 +1,220 @@
+                                match ScriptRuntime::new(&script, AST(timer.clone())) {
+                                    Ok(r) => {
+                                        ret.send(Ok(())).ok();
+                                        runtime = r;
+                                        log::info!(target: "Auto Splitter", "Reloaded script");
+                                    }
+                                    Err(_) => {
+                                        ret.send(Err(Error::LoadFailed)).ok();
+                                        log::info!(target: "Auto Splitter", "Failed to load");
+                                    }
+                                }
+                            }
+                            Request::UnloadScript(ret) => {
+                                ret.send(()).ok();
+                                log::info!(target: "Auto Splitter", "Unloaded script");
+                                continue 'back_to_not_having_a_runtime;
+                            }
+                            Request::End => return Ok(()),
+                        }
+                    }
+                    if let Err(e) = runtime.step() {
+                        log::error!(target: "Auto Splitter", "Unloaded due to failure: {}", e);
+                        continue 'back_to_not_having_a_runtime;
+                    };
+                    runtime.sleep();
+                }
+            }
+        });
+
+        Self {
+            sender,
+            join_handle: Some(join_handle),
+        }
+    }
+
+    /// Attempt to load a wasm file containing an autosplitter module. This call
+    /// will block until the autosplitter has either loaded successfully or
+    /// failed.
+    pub fn load_script(&self, script: PathBuf) -> Result<()> {
+        // TODO: replace with `futures::channel::oneshot`
+        let (sender, receiver) = bounded(1);
+        self.sender
+            .send(Request::LoadScript(script, sender))
+            .map_err(|_| Error::ThreadStopped)?;
+        receiver
+            .recv()
+            .map_err(|_| Error::ThreadStopped)
+            .and_then(std::convert::identity)
+    }
+
+    /// Unload the current autosplitter. This will _not_ return an error if
+    /// there isn't currently an autosplitter loaded, only if the runtime thread
+    /// stops unexpectedly.
+    pub fn unload_script(&self) -> Result<()> {
+        // TODO: replace with `futures:channel::oneshot`
+        let (sender, receiver) = bounded(1);
+        self.sender
+            .send(Request::UnloadScript(sender))
+            .map_err(|_| Error::ThreadStopped)?;
+        receiver.recv().map_err(|_| Error::ThreadStopped)
+    }
+}
+
+enum Request {
+    LoadScript(PathBuf, Sender<Result<()>>),
+    UnloadScript(Sender<()>),
+    End,
+}
+
+// This newtype is required because SharedTimer is an Arc, so we can't impl
+// the autosplitter Timer trait directly on it
+struct AST(SharedTimer);
```

Comment:
Is this even true? Usually you can implement traits if at least part of the type (being the Timer) is defined in your own crate. I might be wrong though. But worth checking again.